hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java

```diff
@@ -98,20 +99,34 @@ private static Artifact checkForExistingArtifact(final String filename, final bo
     public Artifact create(final ArtifactUpload artifactUpload) {
         final String filename = artifactUpload.getFilename();
         final long moduleId = artifactUpload.getModuleId();
-        AbstractDbArtifact result = null;
 
         final SoftwareModule softwareModule = getModuleAndThrowExceptionIfThatFails(moduleId);
 
-        final Artifact existing = checkForExistingArtifact(filename, artifactUpload.overrideExisting(),
-                softwareModule);
+        final Artifact existing = checkForExistingArtifact(filename, artifactUpload.overrideExisting(), softwareModule);
 
         assertArtifactQuota(moduleId, 1);
         assertMaxArtifactSizeQuota(filename, moduleId, artifactUpload.getFilesize());
         assertMaxArtifactStorageQuota(filename, artifactUpload.getFilesize());
 
+        return  getOrCreateArtifact(artifactUpload)
+                .map(artifact -> storeArtifactMetadata(softwareModule, filename, artifact, existing))
+                .orElse(null);
+    }
+
+    private Optional<AbstractDbArtifact> getOrCreateArtifact(final ArtifactUpload artifactUpload) {
+        final String providedSha1Sum = artifactUpload.getProvidedSha1Sum();
+        AbstractDbArtifact artifact = null;
+
+        if (!StringUtils.isEmpty(providedSha1Sum)) {
+            artifact = artifactRepository.getArtifactBySha1(tenantAware.getCurrentTenant(), providedSha1Sum);
+        }
+        return Optional.ofNullable(artifact == null ? storeArtifact(artifactUpload) : artifact);
```

Comment:
I would still propose my previous solution (with 2 return statements and final AbstractDbArtifact) in order to get rid of conditional here

Reply:
Yes, but then we have "multiple return statements" and we still have to do the null check anyway.
I could move the check to an own line to make it simpler maybe. What do you think?
```
artifact = (artifact == null) ? storeArtifact(artifactUpload) : artifact;
return Optional.ofNullable(artifact);
```

Reply:
Ok, let it be our consensus for this question :)